### ISSUE TITLE
PHP 7.3/7.4 failed compiles with multiple definition of `HeapOverflowSanityCheckGetSize' ?

### ISSUE THREAD

**Issue by centminmod** [Imagick/imagick#311]
Seems latest PHP 7.3 and 7.4 are getting errors on imagick 3.4.4 compiles with `multiple definition of HeapOverflowSanityCheckGetSize`

```
/bin/sh /svr-setup/imagick-3.4.4/libtool --mode=link ccache gcc -DPHP_ATOM_INC -I/svr-setup/imagick-3.4.4/include -I/svr-setup/imagick-3.4.4/main -I/svr-setup/imagick-3.4.4 -I/usr/local/include/php -I/usr/local/include/php/main -I/usr/local/include/php/TSRM -I/usr/local/include/php/Zend -I/usr/local/include/php/ext -I/usr/local/include/php/ext/date/lib -I/usr/include/ImageMagick-6  -DHAVE_CONFIG_H  -g -O2    -o imagick.la -export-dynamic -avoid-version -prefer-pic -module -rpath /svr-setup/imagick-3.4.4/modules  imagick_file.lo imagick_class.lo imagickdraw_class.lo imagickpixel_class.lo imagickpixeliterator_class.lo imagick_helpers.lo imagick.lo imagickkernel_class.lo shim_im6_to_im7.lo -lMagickWand-6.Q16 -lMagickCore-6.Q16
ccache gcc -shared  .libs/imagick_file.o .libs/imagick_class.o .libs/imagickdraw_class.o .libs/imagickpixel_class.o .libs/imagickpixeliterator_class.o .libs/imagick_helpers.o .libs/imagick.o .libs/imagickkernel_class.o .libs/shim_im6_to_im7.o  -lMagickWand-6.Q16 -lMagickCore-6.Q16  -Wl,-soname -Wl,imagick.so -o .libs/imagick.so
.libs/imagick_class.o: In function `HeapOverflowSanityCheck':
/usr/include/ImageMagick-6/magick/memory_.h:74: multiple definition of `HeapOverflowSanityCheck'
.libs/imagick_file.o:/usr/include/ImageMagick-6/magick/memory_.h:74: first defined here
.libs/imagick_class.o: In function `HeapOverflowSanityCheckGetSize':
/usr/include/ImageMagick-6/magick/memory_.h:90: multiple definition of `HeapOverflowSanityCheckGetSize'
.libs/imagick_file.o:/usr/include/ImageMagick-6/magick/memory_.h:90: first defined here
.libs/imagickdraw_class.o: In function `HeapOverflowSanityCheck':
/usr/include/ImageMagick-6/magick/memory_.h:74: multiple definition of `HeapOverflowSanityCheck'
.libs/imagick_file.o:/usr/include/ImageMagick-6/magick/memory_.h:74: first defined here
.libs/imagickdraw_class.o: In function `HeapOverflowSanityCheckGetSize':
/usr/include/ImageMagick-6/magick/memory_.h:90: multiple definition of `HeapOverflowSanityCheckGetSize'
.libs/imagick_file.o:/usr/include/ImageMagick-6/magick/memory_.h:90: first defined here
.libs/imagickpixel_class.o: In function `HeapOverflowSanityCheck':
/usr/include/ImageMagick-6/magick/memory_.h:74: multiple definition of `HeapOverflowSanityCheck'
.libs/imagick_file.o:/usr/include/ImageMagick-6/magick/memory_.h:74: first defined here
.libs/imagickpixel_class.o: In function `HeapOverflowSanityCheckGetSize':
/usr/include/ImageMagick-6/magick/memory_.h:90: multiple definition of `HeapOverflowSanityCheckGetSize'
.libs/imagick_file.o:/usr/include/ImageMagick-6/magick/memory_.h:90: first defined here
.libs/imagickpixeliterator_class.o: In function `HeapOverflowSanityCheck':
/usr/include/ImageMagick-6/magick/memory_.h:74: multiple definition of `HeapOverflowSanityCheck'
.libs/imagick_file.o:/usr/include/ImageMagick-6/magick/memory_.h:74: first defined here
.libs/imagickpixeliterator_class.o: In function `HeapOverflowSanityCheckGetSize':
/usr/include/ImageMagick-6/magick/memory_.h:90: multiple definition of `HeapOverflowSanityCheckGetSize'
.libs/imagick_file.o:/usr/include/ImageMagick-6/magick/memory_.h:90: first defined here
.libs/imagick_helpers.o: In function `HeapOverflowSanityCheck':
/usr/include/ImageMagick-6/magick/memory_.h:74: multiple definition of `HeapOverflowSanityCheck'
.libs/imagick_file.o:/usr/include/ImageMagick-6/magick/memory_.h:74: first defined here
.libs/imagick_helpers.o: In function `HeapOverflowSanityCheckGetSize':
/usr/include/ImageMagick-6/magick/memory_.h:90: multiple definition of `HeapOverflowSanityCheckGetSize'
.libs/imagick_file.o:/usr/include/ImageMagick-6/magick/memory_.h:90: first defined here
.libs/imagick.o: In function `HeapOverflowSanityCheck':
/usr/include/ImageMagick-6/magick/memory_.h:74: multiple definition of `HeapOverflowSanityCheck'
.libs/imagick_file.o:/usr/include/ImageMagick-6/magick/memory_.h:74: first defined here
.libs/imagick.o: In function `HeapOverflowSanityCheckGetSize':
/usr/include/ImageMagick-6/magick/memory_.h:90: multiple definition of `HeapOverflowSanityCheckGetSize'
.libs/imagick_file.o:/usr/include/ImageMagick-6/magick/memory_.h:90: first defined here
.libs/imagickkernel_class.o: In function `HeapOverflowSanityCheck':
/usr/include/ImageMagick-6/magick/memory_.h:74: multiple definition of `HeapOverflowSanityCheck'
.libs/imagick_file.o:/usr/include/ImageMagick-6/magick/memory_.h:74: first defined here
.libs/imagickkernel_class.o: In function `HeapOverflowSanityCheckGetSize':
/usr/include/ImageMagick-6/magick/memory_.h:90: multiple definition of `HeapOverflowSanityCheckGetSize'
.libs/imagick_file.o:/usr/include/ImageMagick-6/magick/memory_.h:90: first defined here
.libs/shim_im6_to_im7.o: In function `HeapOverflowSanityCheck':
/usr/include/ImageMagick-6/magick/memory_.h:74: multiple definition of `HeapOverflowSanityCheck'
.libs/imagick_file.o:/usr/include/ImageMagick-6/magick/memory_.h:74: first defined here
.libs/shim_im6_to_im7.o: In function `HeapOverflowSanityCheckGetSize':
/usr/include/ImageMagick-6/magick/memory_.h:90: multiple definition of `HeapOverflowSanityCheckGetSize'
.libs/imagick_file.o:/usr/include/ImageMagick-6/magick/memory_.h:90: first defined here
collect2: error: ld returned 1 exit status
make: *** [imagick.la] Error 1
```

The last working imagick build had the following on CentOS 7.7 64bit using Remi ImageMagick yum 

```
yum list installed ImageMagick6* -q | tr -s ' ' | column -t
Installed                      Packages
ImageMagick6.x86_64            6.9.10.79-1.el7.remi  @remi
ImageMagick6-c++.x86_64        6.9.10.79-1.el7.remi  @remi
ImageMagick6-c++-devel.x86_64  6.9.10.79-1.el7.remi  @remi
ImageMagick6-devel.x86_64      6.9.10.79-1.el7.remi  @remi
ImageMagick6-libs.x86_64       6.9.10.79-1.el7.remi  @remi
```
```
php --ri imagick

imagick

imagick module => enabled
imagick module version => 3.4.4
imagick classes => Imagick, ImagickDraw, ImagickPixel, ImagickPixelIterator, ImagickKernel
Imagick compiled with ImageMagick version => ImageMagick 6.9.10-79 Q16 x86_64 2019-12-18 https://imagemagick.org
Imagick using ImageMagick library version => ImageMagick 6.9.10-79 Q16 x86_64 2019-12-18 https://imagemagick.org
ImageMagick copyright => © 1999-2020 ImageMagick Studio LLC
ImageMagick release date => 2019-12-18
ImageMagick number of supported formats:  => 242
ImageMagick supported formats => 3FR, 3G2, 3GP, AAI, AI, ART, ARW, AVI, AVS, BGR, BGRA, BGRO, BIE, BMP, BMP2, BMP3, BRF, CAL, CALS, CANVAS, CAPTION, CIN, CIP, CLIP, CMYK, CMYKA, CR2, CR3, CRW, CUR, CUT, DATA, DCM, DCR, DCX, DDS, DFONT, DNG, DOT, DPX, DXT1, DXT5, EPDF, EPI, EPS, EPS2, EPS3, EPSF, EPSI, EPT, EPT2, EPT3, ERF, EXR, FAX, FILE, FITS, FRACTAL, FTP, FTS, G3, G4, GIF, GIF87, GRADIENT, GRAY, GRAYA, GROUP4, GV, H, HALD, HDR, HISTOGRAM, HRZ, HTM, HTML, HTTP, HTTPS, ICB, ICO, ICON, IIQ, INFO, INLINE, IPL, ISOBRL, ISOBRL6, J2C, J2K, JBG, JBIG, JNG, JNX, JP2, JPC, JPE, JPEG, JPG, JPM, JPS, JPT, JSON, K25, KDC, LABEL, M2V, M4V, MAC, MAGICK, MAP, MASK, MAT, MATTE, MEF, MIFF, MKV, MNG, MONO, MOV, MP4, MPC, MPEG, MPG, MRW, MSL, MSVG, MTV, MVG, NEF, NRW, NULL, ORF, OTB, OTF, PAL, PALM, PAM, PANGO, PATTERN, PBM, PCD, PCDS, PCL, PCT, PCX, PDB, PDF, PDFA, PEF, PES, PFA, PFB, PFM, PGM, PGX, PICON, PICT, PIX, PJPEG, PLASMA, PNG, PNG00, PNG24, PNG32, PNG48, PNG64, PNG8, PNM, POCKETMOD, PPM, PREVIEW, PS, PS2, PS3, PSB, PSD, PTIF, PWP, RADIAL-GRADIENT, RAF, RAS, RAW, RGB, RGBA, RGBO, RGF, RLA, RLE, RMF, RW2, SCR, SCT, SFW, SGI, SHTML, SIX, SIXEL, SPARSE-COLOR, SR2, SRF, STEGANO, SUN, SVG, SVGZ, TEXT, TGA, THUMBNAIL, TIFF, TIFF64, TILE, TIM, TTC, TTF, TXT, UBRL, UBRL6, UIL, UYVY, VDA, VICAR, VID, VIFF, VIPS, VST, WBMP, WEBP, WMF, WMV, WMZ, WPG, X, X3F, XBM, XC, XCF, XPM, XPS, XV, XWD, YCbCr, YCbCrA, YUV

Directive => Local Value => Master Value
imagick.locale_fix => 0 => 0
imagick.skip_version_check => 0 => 0
imagick.progress_monitor => 0 => 0
```
PHP build
```
php-config
Usage: /usr/local/bin/php-config [OPTION]
Options:
  --prefix            [/usr/local]
  --includes          [-I/usr/local/include/php -I/usr/local/include/php/main -I/usr/local/include/php/TSRM -I/usr/local/include/php/Zend -I/usr/local/include/php/ext -I/usr/local/include/php/ext/date/lib]
  --ldflags           [ -L/usr/local/lib64]
  --libs              [-lcrypt  -lc-client  -ltidy -largon2 -lresolv -lcrypt -lncurses -laspell -lpspell -lrt -lldap -llber -lstdc++ -lcrypt -lpam -lgmp -lbz2 -lrt -lm -ldl  -lsystemd -lxml2 -lgssapi_krb5 -lkrb5 -lk5crypto -lcom_err -lssl -lcrypto -lsqlite3 -lz -lcurl -lxml2 -lenchant -lgmodule-2.0 -lglib-2.0 -lffi -lssl -lcrypto -lz -lpng15 -lwebp -ljpeg -lXpm -lX11 -lfreetype -lgssapi_krb5 -lkrb5 -lk5crypto -lcom_err -lssl -lcrypto -licuio -licui18n -licuuc -licudata -lonig -lsqlite3 -ledit -ltinfo -lxml2 -lnetsnmp -lssl -lssl -lcrypto -lm -lxml2 -lsodium -lcrypt -largon2 -lxml2 -lxml2 -lxml2 -lxml2 -lxslt -lm -lxml2 -lexslt -lxslt -lm -lgcrypt -ldl -lgpg-error -lxml2 -lzip -lz -lssl -lcrypto -lcrypt ]
  --extension-dir     [/usr/local/lib/php/extensions/no-debug-non-zts-20190902]
  --include-dir       [/usr/local/include/php]
  --man-dir           [/usr/local/php/man]
  --php-binary        [/usr/local/bin/php]
  --php-sapis         [ cli embed fpm phpdbg cgi]
  --ini-path          [/usr/local/lib]
  --ini-dir           [/etc/centminmod/php.d]
  --configure-options [--enable-fpm --enable-opcache --enable-intl --enable-pcntl --with-mcrypt --with-snmp --enable-embed=shared --with-mhash --with-zlib --with-gettext --enable-exif --with-zip --with-libzip --with-bz2 --enable-soap --enable-sockets --enable-sysvmsg --enable-sysvsem --enable-sysvshm --with-mysql-sock=/var/lib/mysql/mysql.sock --with-curl --enable-gd --with-xmlrpc --enable-bcmath --enable-calendar --enable-ftp --enable-gd-native-ttf --with-freetype --with-jpeg --with-png-dir=/usr --with-xpm --with-webp --with-t1lib=/usr --enable-shmop --with-pear --enable-mbstring --with-openssl --with-mysql=mysqlnd --with-libdir=lib64 --with-mysqli=mysqlnd --enable-pdo --with-pdo-sqlite --with-pdo-mysql=mysqlnd --enable-inline-optimization --with-imap --with-imap-ssl --with-kerberos --with-readline --with-libedit --with-gmp --with-pspell --with-tidy --with-enchant --with-fpm-user=nginx --with-fpm-group=nginx --with-ldap --with-ldap-sasl --with-password-argon2=/usr/local --with-sodium=/usr/local --with-config-file-scan-dir=/etc/centminmod/php.d --with-fpm-systemd --with-ffi --with-xsl PKG_CONFIG_PATH=/opt/rh/devtoolset-8/root/usr/lib64/pkgconfig LIBSODIUM_CFLAGS=-I/usr/local/include LIBSODIUM_LIBS=-L/usr/local/lib64 -lsodium LIBZIP_CFLAGS=-I/usr/local/include LIBZIP_LIBS=-L/usr/local/lib64 -lzip]
  --version           [7.4.1]
  --vernum            [70401]
```

**Comment by Danack** [comment]
Please can you do a full clean build of Imagick, in a new directory if needed.

It looks like somehow you're linking against two separate versions of ImageMagick.

```
`HeapOverflowSanityCheck' .libs/imagick_file.o:/usr/include/ImageMagick-6/magick/memory_.h:74: first defined here
`HeapOverflowSanityCheckGetSize'
.libs/imagick_file.o:/usr/include/ImageMagick-6/magick/memory_.h:90: first defined here
```

The multiple definition it is complaining about isn't called directly from Imagick.....so it's unlikely to be something in the code from this repo.

If doing that doesn't work, please do a full rebuild against a released version from https://github.com/ImageMagick/ImageMagick6 - if that still doesn't work you probably should investigate why there is a double definition, and then open an issue in that repo.

If that version does work, you should probably open a bug against the repo for the version of ImageMagick you're using.

But I strongly suspect doing a full clean build will make this issue go away.

**Comment by centminmod** [comment]
Thanks for the hints, I am building against Remi CentOS 7's ImageMagick6 repo

```
rpm -ql ImageMagick6-devel | grep 'memory_.h'  
/usr/include/ImageMagick-6/magick/memory_.h
```
```
rpm -qa ImageMagick6*
ImageMagick6-6.9.10.79-1.el7.remi.x86_64
ImageMagick6-c++-devel-6.9.10.79-1.el7.remi.x86_64
ImageMagick6-devel-6.9.10.79-1.el7.remi.x86_64
ImageMagick6-libs-6.9.10.79-1.el7.remi.x86_64
ImageMagick6-c++-6.9.10.79-1.el7.remi.x86_64
```

will have to figure out why it suddenly failing when it previously worked

looks like there were some yum updates recently

```
yum history list ImageMagick6-devel
Loaded plugins: fastestmirror, priorities, versionlock
ID     | Command line             | Date and time    | Action(s)      | Altered
-------------------------------------------------------------------------------
   601 | -y update ImageMagick6 I | 2019-12-18 14:28 | Update         |    5   
   600 | -y update ImageMagick6 I | 2019-12-12 06:10 | Update         |    5   
   594 | -y update ImageMagick6 I | 2019-12-04 01:29 | Update         |    5   
   591 | -y update ImageMagick6 I | 2019-11-27 17:40 | Update         |    5   
```

**Comment by Danack** [comment]
Possibly related, but I still think it looks like a 'not cleaning everything' problem.

The function it is complaining about is only listed once in that file, but the error message refers to two separate lines:

```
memory_.h:74: first defined here
memory_.h:90: first defined here
```

The function it's complaining about is only listed once in that file, so the only possible way it would be included twice would be for multiple versions, right?

Remi has pretty good test coverage...and he usually finds these types of things quite quickly, so I still think it's something needing cleaning on your system.

**Comment by centminmod** [comment]
cheers @Danack will investigate :)

**Comment by kleisauke** [comment]
fwiw, I ran into the same linker errors while cross-compiling ImageMagick 6.9.10-80 with MinGW-w64: https://gist.github.com/kleisauke/d385c6b02d31690fb9b22a6ddecad612

This patch fixes it for me:
```diff
diff --git a/magick/memory_.h b/magick/memory_.h
index 1111111..2222222 100644
--- a/magick/memory_.h
+++ b/magick/memory_.h
@@ -68,7 +68,7 @@ extern MagickExport void
   SetMagickMemoryMethods(AcquireMemoryHandler,ResizeMemoryHandler,
     DestroyMemoryHandler);
 
-inline MagickExport MagickBooleanType HeapOverflowSanityCheck(
+inline MagickBooleanType HeapOverflowSanityCheck(
   const size_t count,const size_t quantum)
 {
   if ((count == 0) || (quantum == 0))
@@ -81,7 +81,7 @@ inline MagickExport MagickBooleanType HeapOverflowSanityCheck(
   return(MagickFalse);
 }
 
-inline MagickExport MagickBooleanType HeapOverflowSanityCheckGetSize(
+inline MagickBooleanType HeapOverflowSanityCheckGetSize(
   const size_t count,const size_t quantum,size_t *const extent)
 {
   size_t
````

/cc @remicollet because he might also be interested in this issue and the patch above.

**Comment by pamamolf** [comment]
Yes this patch is fixing the issue.

It will be great if added !

Thank you

**Comment by Danack** [comment]
@kleisauke maybe submit that as a PR at https://github.com/ImageMagick/ImageMagick ?

**Comment by pamamolf** [comment]
Yes please submit it as a pull request !

**Comment by kleisauke** [comment]
These linker errors should be fixed in ImageMagick 6.9.10-82 and 7.0.9-12, which has just been released on Remi's RPM repository.

**Comment by Danack** [comment]
Fixed upstream (probably, haven't double checked), so closing.
